iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp

```diff
@@ -209,6 +244,307 @@ class GlobalStoreOpConversion : public OpConversionPattern<StoreOpTy> {
   StringRef funcName;
 };
 
+// Convert vm list operations to two emitc calls. The wrapping ref pointer is
+// first dereferenced and the result is used as the argument of the specified
+// function name.
+template <typename SrcOpTy>
+class ListOpConversion : public OpConversionPattern<SrcOpTy> {
+  using OpConversionPattern<SrcOpTy>::OpConversionPattern;
+
+ public:
+  ListOpConversion(MLIRContext *context, StringRef funcName,
+                   size_t listArgumentIndex)
+      : OpConversionPattern<SrcOpTy>(context),
+        funcName(funcName),
+        listArgumentIndex(listArgumentIndex) {}
+
+ private:
+  LogicalResult matchAndRewrite(
+      SrcOpTy op, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const override {
+    auto ctx = op.getContext();
+    auto loc = op.getLoc();
+
+    if (listArgumentIndex >= operands.size()) {
+      return op.emitError() << " index for list argument out of range";
+    }
+
+    Value listOperand = op.getOperation()->getOperand(listArgumentIndex);
+
+    auto listDerefOp = rewriter.create<emitc::CallOp>(
+        /*location=*/loc,
+        /*type=*/emitc::OpaqueType::get(ctx, "iree_vm_list_t*"),
+        /*callee=*/rewriter.getStringAttr("iree_vm_list_deref"),
+        /*args=*/ArrayAttr{},
+        /*templateArgs=*/ArrayAttr{},
+        /*operands=*/ArrayRef<Value>{listOperand});
+
+    // Replace the one list argument (which is wrapped in a ref) with the
+    // unwrapped list.
+    SmallVector<Value, 4> updatedOperands;
+    for (auto &operand : llvm::enumerate(operands)) {
+      if (operand.index() == listArgumentIndex) {
+        updatedOperands.push_back(listDerefOp.getResult(0));
+      } else {
+        updatedOperands.push_back(operand.value());
+      }
+    }
+
+    rewriter.replaceOpWithNewOp<emitc::CallOp>(
+        /*op=*/op,
+        /*type=*/op.getOperation()->getResultTypes(),
+        /*callee=*/rewriter.getStringAttr(funcName),
+        /*args=*/ArrayAttr{},
+        /*templateArgs=*/ArrayAttr{},
+        /*operands=*/ArrayRef<Value>(updatedOperands));
+
+    return releaseRefIfLastUse(op.getOperation(), listOperand);
+  }
+
+  StringRef funcName;
+
+  // The index of the list argument. This gets replaced in the conversion.
+  size_t listArgumentIndex;
+};
+
+class ListAllocOpConversion
+    : public OpConversionPattern<IREE::VM::ListAllocOp> {
+  using OpConversionPattern<IREE::VM::ListAllocOp>::OpConversionPattern;
+
+ private:
+  LogicalResult matchAndRewrite(
+      IREE::VM::ListAllocOp allocOp, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const override {
+    // clang-format off
+    // The generated c code looks roughly like this.
+    // iree_vm_type_def_t element_type = iree_vm_type_def_make_value_type(IREE_VM_VALUE_TYPE_I32);
+    // iree_vm_type_def_t* element_type_ptr = &element_type;
```

Comment:
is this type_ptr/list_ptr needed?

Reply:
Yes these are used in the calls below actually. I'll update the comment.